tests/e2e/support/mock_client.rs

```diff
@@ -440,61 +447,90 @@ impl MockClient {
+    /// AnalysisFinished get emitted via `cairo/serverStatus` notification).
+    pub fn open_absolute_path_and_wait_for_diagnostics_generation(
+        &mut self,
+        path: impl AsRef<Path>,
+    ) -> HashMap<Url, Vec<Diagnostic>> {
+        self.open_absolute_path(path);
 
         let mut in_progress_open = false;
         let mut project_updated = false;
 
-        self.subscribe_notifications(Box::new(move |notification: Notification| {
-            if notification.method == "cairo/serverStatus" {
-                let params: ServerStatusParams =
-                    serde_json::from_value(notification.params).unwrap();
+        self.subscribe_notifications(Box::new(
+            move |notification: Notification| match notification.method.as_str() {
+                "cairo/serverStatus" => {
+                    let params: ServerStatusParams =
+                        serde_json::from_value(notification.params).unwrap();
 
-                match params.event {
-                    AnalysisStarted => {
-                        in_progress_open = true;
-                    }
-                    AnalysisFinished => {
-                        if in_progress_open && project_updated {
-                            return ControlFlow::Break(());
+                    match params.event {
+                        AnalysisStarted => {
+                            in_progress_open = true;
+                            ControlFlow::Continue(true)
+                        }
+                        AnalysisFinished => {
+                            if in_progress_open && project_updated {
+                                ControlFlow::Break(())
+                            } else {
+                                ControlFlow::Continue(true)
+                            }
                         }
                     }
                 }
-            }
-            if notification.method == "cairo/projectUpdatingFinished" {
-                project_updated = true;
-            }
-
-            ControlFlow::Continue(())
-        }));
+                "cairo/projectUpdatingFinished" => {
+                    project_updated = true;
+                    ControlFlow::Continue(true)
+                }
+                _ => ControlFlow::Continue(false),
+            },
+        ));
 
         // It should contain all accumulated diagnostics up to the end of the generation.
         self.diagnostics.clone()
     }
 
     fn subscribe_notifications(&mut self, mut stream_consumer: StreamConsumer<Notification>) {
-        for msg in self.trace() {
+        for (i, msg) in self.trace.clone().into_iter().enumerate() {
             if let Message::Notification(notification) = msg {
-                if stream_consumer(notification.clone()).is_break() {
-                    return;
-                }
+                match stream_consumer(notification) {
+                    ControlFlow::Continue(true) => {
+                        self.trace.remove(i);
```

Comment:
This will be wrong when you operate on cloned version, whilst the original one changes length during this sequence of operations on stream. 
You will have wrong indices, possibly even out of bound ones